engine.fhir/src/main/java/org/opencds/cqf/cql/engine/fhir/retrieve/BaseFhirQueryGenerator.java

```diff
@@ -0,0 +1,287 @@
+import org.opencds.cqf.cql.engine.runtime.DateTime;
+import org.opencds.cqf.cql.engine.runtime.Interval;
+import org.opencds.cqf.cql.engine.terminology.TerminologyProvider;
+import org.opencds.cqf.cql.engine.terminology.ValueSetInfo;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.context.RuntimeSearchParam;
+import ca.uhn.fhir.model.api.IQueryParameterType;
+import ca.uhn.fhir.rest.api.RestSearchParameterTypeEnum;
+import ca.uhn.fhir.rest.param.DateParam;
+import ca.uhn.fhir.rest.param.DateRangeParam;
+import ca.uhn.fhir.rest.param.ParamPrefixEnum;
+import ca.uhn.fhir.rest.param.TokenOrListParam;
+import ca.uhn.fhir.rest.param.TokenParam;
+import ca.uhn.fhir.rest.param.TokenParamModifier;
+
+public abstract class BaseFhirQueryGenerator {
+    protected static final int DEFAULT_MAX_CODES_PER_QUERY = 64;
+    protected static final boolean DEFAULT_SHOULD_EXPAND_VALUESETS = false;
+
+    protected FhirContext fhirContext;
+
+    protected TerminologyProvider terminologyProvider;
+    protected SearchParameterResolver searchParameterResolver;
+    protected ModelResolver modelResolver;
+
+    private Integer pageSize;
+    public Integer getPageSize() {
+        return this.pageSize;
+    }
+    public void setPageSize(Integer value) {
+        if( value == null || value < 1 ) {
+            throw new IllegalArgumentException("value must be a non-null integer > 0");
+        }
+        this.pageSize = value;
+    }
+
+    private int maxCodesPerQuery;
+    public int getMaxCodesPerQuery() {
+        return this.maxCodesPerQuery;
+    }
+    public void setMaxCodesPerQuery(int value) {
+        if (value < 1) {
+            throw new IllegalArgumentException("value must be > 0");
+        }
+        this.maxCodesPerQuery = value;
+    }
+
+    // TODO: Think about how to best handle the decision to expand value sets... Should it be part of the
+    // terminology provider if it detects support for "code:in"? How does that feed back to the retriever?
+    private boolean expandValueSets;
+    public boolean isExpandValueSets() {
+        return this.expandValueSets;
+    }
+    public void setExpandValueSets(boolean expandValueSets) {
+        this.expandValueSets = expandValueSets;
+    }
+
+    protected BaseFhirQueryGenerator(SearchParameterResolver searchParameterResolver, TerminologyProvider terminologyProvider,
+                                  ModelResolver modelResolver, FhirContext fhirContext) {
+        this.searchParameterResolver = searchParameterResolver;
+        this.terminologyProvider = terminologyProvider;
+        this.modelResolver = modelResolver;
+        this.maxCodesPerQuery = DEFAULT_MAX_CODES_PER_QUERY;
+        this.expandValueSets = DEFAULT_SHOULD_EXPAND_VALUESETS;
+
+        this.fhirContext = fhirContext;
+    }
+
+    public abstract List<String> generateFhirQueries(ICompositeType dataRequirement, Context engineContext, IBaseConformance capabilityStatement);
```

Comment:
We can't pass in engine context here. These pieces are built prior to engine context existing.

Reply:
If you need specific context for doing query generation, create a class or holder object to store those. Put another way, if we need a full cql-engine to do query generation we need to refactor some things.